### DISCUSSION TITLE
ACCESSING DATA STRUCTURES BETWEEN TWO OR THREE DIFFERENT REACH PROGRAMS

### DISCUSSION THREAD

**Discussion by kcelestinomaria** [reach-sh/reach-lang#1199] [Ideas]
Hello Reach Fam,

Say I have two smart contracts(Main Reach Programs), ```Reach Program A``` has ```Participant.Alice``` who ```publishes()``` some ```info```, this ```info``` is added as an entry in a new Map, ```MapInfo``` . I then have ```Reach Program B``` which has ```Participant.John```,  who needs to add a new entry and change some info in Alice's published info. The question is, how can ```Reach Program B``` access MapInfo from Reach Program A, and change specific entries in MapInfo by adding new info provided by  ```Participant.Bob``` ?

In short, is there a simple way for a Reach Program to access another Reach Program's internal data structures without much restriction and still following Reach's rules? Also, the examples on the Reach Official Documentation only show how to call functions in remote programs, but we are simply trying to change entries in a Map between two different Reach Programs in this thought problem. Thanks.

**Comment by jeapostrophe** [comment]
The only thing that can change A's Map is A. So you need to have B call a function of A that changes it. This is just like normal programming where you define interfaces that do work for clients.

**Comment by kcelestinomaria** [comment]
Hello @jeapostrophe , does this work for functions defined in ```Views``` as well?

**Reply by jeapostrophe** [reply]
Yup, https://docs.reach.sh/rsh/appinit/#p_44

**Reply by kcelestinomaria** [reply]
And, how do functions successfully return Maps when called? I did this:

```
const MaybeMap = Maybe(Map);

const User = Participant('User', { outputAssetMap: Fun([MAP_KEY_TYPE], MaybeMap) }

const MAP_KEY_TYPE = Bytes(20);

const assetMap = new Map(MAP_KEY_TYPE, assetState); // assetState is some struct

//create a new Map entry with some data published during consensus
assetMap[someCIDHash] = assetState.fromObject({ assetName: pbAssetName, pbAssetValue: pbAssetValue });

// State is a View that reads state of assets from the contract
State.readAsset.set((someCIDHash) => assetMap[someCIDHash]);

// defined outputAssetMap method as outputAssetMap([MAP_KEY_TYPE], MaybeMap)
User.interact.outputAssetMap([someCIDHash], assetMap[someCIDHash]);
```
Does the above method work for calling specific Map entries?
My goal is that I then want to call ```outputAssetMap()``` in another contract, receive a specific Map entry by ```someCIDHash```, and then allow participants in this other contract to change the one of the Map's values with the specific ```someCIDHash``` that is the Map's Keys(this variable is voluntarily provided by the participants, and they expect a Map's value output) @jeapostrophe

**Reply by jeapostrophe** [reply]
You cannot return an entire Map. You'll have to provide a function that accesses one entry at a time

**Reply by kcelestinomaria** [reply]
> You cannot return an entire Map. You'll have to provide a function that accesses one entry at a time
Which one returns the specific entry?
```Fun([MAP_KEY_TYPE], assetMap[MAP_KEY_TYPE]) ``` 
OR
```Fun([MAP_KEY_TYPE], assetMap)```
OR
```Fun([MAP_KEY_TYPE], MaybeMap)``` @jeapostrophe , cause so far all examples show the most straightforward function call returns another data structure as the value, that is:
```Fun([MAP_KEY_TYPE], assetState)``` but the goal here is to return that specific Map entry

**Reply by jeapostrophe** [reply]
Suppose you have a Map...

```
const m = new Map(UInt)
```

then you could have a function

```
const p = View({
 getM: Fun([Address], Maybe(UInt)
})
```

and an implementation

```
p.getM.set((a) => m[a])
```
